lib/Base/System/ClientLists.vala

```diff
@@ -21,290 +21,169 @@
- *
- * Can be used by an application to retrieve an updating list of currently
- * active servers or users, and is used internally to keep track of them.
+ * Provides a list of all sessions used by a Client.
  */
-public abstract class Backend.ClientList <T> : ListModel, Object {
-
-  /**
-   * Runs at construction of a new instance.
-   */
-  construct {
-    store = new GenericArray <T> ();
-  }
-
-  /**
-   * Returns the nth session in this list.
-   *
-   * @param position The position to look for.
-   *
-   * @return The session at the position, or null if position is invalid.
-   */
-  public Object? get_item (uint position) {
-    return store.get (position) as Object;
-  }
-
-  /**
-   * Get the number of sessions in this list.
-   *
-   * @return The number of sessions in this list.
-   */
-  public uint get_n_items () {
-    return store.length;
-  }
-
-  /**
-   * Returns the type of the objects this ListModel stores.
-   *
-   * @return The type for a base Session class.
-   */
-  public Type get_item_type () {
-    return typeof (T);
-  }
-
-  /**
-   * Check if a item exists with a specific search function.
-   *
-   * @param needle The term to search for in the list.
-   * @param equal_func A function used to determine the right item.
-   *
-   * @return The right item if found, else null.
-   */
-  internal T? find <G> (G needle, ArraySearchFunc<T, G> equal_func) {
-    uint? index;
-    if (store.find_custom <G> (needle, equal_func, out index)) {
-      return store.get (index);
-    } else {
-      return null;
-    }
-  }
+public class Backend.SessionList : Backend.Collection<Session> {
 
   /**
-   * Checks if a object is found in the list.
-   *
-   * @param object The object to check for.
-   * @param index The position of the object in the list.
-   *
-   * @return If the object can be found in the list.
+   * Creates a new instance of SessionList.
    */
-  internal bool find_object (T object, out uint index) {
-    return store.find (object, out index);
+  internal SessionList () {
+    Object ();
   }
 
   /**
-   * Adds an item to the list and the
-   * associated access to the KeyStorage.
+   * Add a new session to the session list.
    *
-   * @param item The item to be added.
+   * @param session The session to be added.
    *
-   * @throws Error Errors when adding the access token doesn't work.
+   * @throws Error Errors when saving the secrets to the KeyStorage.
    */
-  internal void add (T item) throws Error {
-    // Stop if item is already in list
-    if (store.find (item)) {
+  internal void add_session (Session session) throws Error {
+    // Avoid duplicates
+    if (find (session, null)) {
       return;
     }
 
-    // Add the item to the list
-    store.add (item);
-
+    // Save session secrets
     try {
-      add_access (item);
+      PlatformEnum platform = PlatformEnum.for_session (session);
+      string token_label = @"Access Token for Account \"$(session.account.username)\" on $(platform)";
```

Comment:
While you're editing this… do we need it to be "username@server" in here? Because if I'm `@ibboard` on two Mastodon servers then these can't be differentiated.

Reply:
I guess that's a good point to make.